kbfsedits/tlf_history.go

```diff
@@ -0,0 +1,362 @@
+	wbr[i], wbr[j] = wbr[j], wbr[i]
+}
+
+func (wbr *writersByRevision) Push(x interface{}) {
+	wn := x.(*writerNotifications)
+	*wbr = append(*wbr, wn)
+}
+
+func (wbr *writersByRevision) Pop() interface{} {
+	// The item to remove is the last item; heap has already swapped
+	// it to the end.
+	old := *wbr
+	n := len(old)
+	item := old[n-1]
+	*wbr = old[0 : n-1]
+	return item
+}
+
+// TlfHistory maintains a history of the last N file edits from each
+// writer in the TLF.
+//
+// There will be two users of a TlfHistory instance:
+//
+//   * One user (likely something outside of the kbfsedits package,
+//     e.g. libkbfs.folderBranchOps) will read notifications from the
+//     corresponding TLF and add them to this history.  After adding a
+//     batch or several batches of messages, it should call
+//     `Recompute()`, and if some writers need more, earlier revisions,
+//     it should fetch more notifications for the indicated writer and
+//     repeat.
+//
+//   * The other user (within the kbfsedits package) will collate the
+//     histories from multiple TlfHistory instances together using
+//     `getHistory()` from each one.  It may also construct pretty
+//     versions of individual edit histories for a particular TLF.
+type TlfHistory struct {
+	lock          sync.RWMutex
+	byWriter      map[string]*writerNotifications
+	computed      bool
+	cachedHistory writersByRevision
+}
+
+// NewTlfHistory constructs a new TlfHistory instance.
+func NewTlfHistory() *TlfHistory {
+	return &TlfHistory{
+		byWriter: make(map[string]*writerNotifications),
+	}
+}
+
+// AddNotifications takes in a set of messages in this TLF by
+// `writer`, and adds them to the history.  Once done adding messages,
+// the caller should call `Recompute` to find out if more messages
+// should be added for any particular writer.
+func (th *TlfHistory) AddNotifications(
+	writerName string, messages []string) (err error) {
+	newEdits := make(notificationsByRevision, 0, len(messages))
+
+	// Unmarshal and sort the new messages.
+	for _, msg := range messages {
+		var revList []NotificationMessage
+		err := json.Unmarshal([]byte(msg), &revList)
+		if err != nil {
+			// The messages might be from a new version we don't
+			// understand, so swallow the error.
+			continue
+		}
+
+		for j := len(revList) - 1; j >= 0; j-- {
+			revMsg := revList[j]
+			if revMsg.Version != NotificationV2 {
+				// Ignore messages that are too new for us to understand.
+				continue
+			}
+			revMsg.numWithinRevision = j
+			newEdits = append(newEdits, revMsg)
+		}
+	}
+
+	th.lock.Lock()
+	defer th.lock.Unlock()
+	wn, existed := th.byWriter[writerName]
+	if !existed {
+		wn = &writerNotifications{writerName, nil}
+	}
+	oldLen := len(wn.notifications)
+	newEdits = append(newEdits, wn.notifications...)
+	sort.Sort(newEdits)
+	wn.notifications = newEdits.uniquify()
+	if len(wn.notifications) == oldLen {
+		// No new messages.
+		return nil
+	}
+	if !existed {
+		wn.writerName = writerName
```

Comment:
`writerName` is already populated above right?

Reply:
I must have had a reason but now I can't remember it, and the tests pass without it, so I'll fix in a new PR.  Thanks for giving it another look!